util/jemalloc_nodump_allocator.cc

```diff
@@ -6,26 +6,98 @@
 #include "util/jemalloc_nodump_allocator.h"
 
 #include <string>
+#include <thread>
 
+#include "port/port.h"
 #include "util/string_util.h"
 
 namespace rocksdb {
+namespace jemalloc {
 
 #ifdef ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
 
 std::atomic<extent_alloc_t*> JemallocNodumpAllocator::original_alloc_{nullptr};
 
 JemallocNodumpAllocator::JemallocNodumpAllocator(
-    unsigned arena_index, int flags, std::unique_ptr<extent_hooks_t>&& hooks)
-    : arena_index_(arena_index), flags_(flags), hooks_(std::move(hooks)) {
-  assert(arena_index != 0);
+    PerCPUArena per_cpu_arena, bool enable_tcache, unsigned num_cpus,
+    std::vector<std::unique_ptr<extent_hooks_t>>&& arena_hooks,
+    std::vector<unsigned>&& arena_indices)
+    : per_cpu_arena_(per_cpu_arena),
+      enable_tcache_(enable_tcache),
+      num_cpus_(num_cpus),
+      arena_hooks_(std::move(arena_hooks)),
```

Comment:
@interwq Shall I hold one `extent_hooks_t` for each arena I create, or shall the arenas shadow the same `extent_hooks_t` struct?

Reply:
Both are fine. Only need to make sure the lifetime of the hooks covers the arenas. If you don't plan to have different hooks, I'd say having one shared struct.